test/type/protocol_types.swift

```diff
@@ -138,3 +138,54 @@ struct BadSubscript {
     set {}
   }
 }
+
+
+protocol TrivialProto {}
```

Comment:
Suggestions for more tests:

- Make sure that nested types of existentials work in both type position (eg, the type of a parameter) and expression position (eg, print(Foo.Bar.self)).

- Add a SILGen test file to make sure SILGen generates things correctly (take the subset of your 'valid' test cases from test/type/protocol_types.swift and put them in a SILGen test).

There's also a fully edge case involving Self:

```
protocol P {
  associatedtype A

  func foo() -> A
}

protocol Q : P where A == Self {}
```

it should be possible to call `foo()` on an existential of type Q.

Reply:
> There's also a fully edge case involving Self

Yeah, `Self` is more tricky. We need a function that recursively searches for a certain associated type in covariant position, in every member up till the protocol that introduces the associated type. We could probably adapt `findProtocolSelfReferences`. But the worst part is we would have to sometimes repeat this for compositions, i.e. for `P1` in `P1(A) & P2(A == Self)`. And to do that we have to propagate the knowledge that A == **Self** together with the map. Maybe by replacing the map's value type with an enum (Opaque, Concrete, Self).

Reply:
> Add a SILGen test file to make sure SILGen generates things correctly (take the subset of your 'valid' test cases from test/type/protocol_types.swift and put them in a SILGen test).

Do I simply paste it in? Unfortunately my knowledge on SILGen is yet superficial, so most the tests look cryptic to me.